From 0773f37d5c93e1bb1f83fb788bc77e752ce8b17b Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Tue, 24 Jul 2018 13:21:22 +0900
Subject: [PATCH 356/909] media: rcar-vin: Add INF bit check for R-Car Gen3

commit 9566489f88efd6cfc1e9d77d15f685a62a50a3d3 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

According to the latest H/W manual, possible setting value of
INF bit for Digital Pins and MIPI-CSI2 are different.

So this patch adds check of INF bit in VnMC register.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/rcar-vin/rcar-dma.c | 24 ++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/media/platform/rcar-vin/rcar-dma.c b/drivers/media/platform/rcar-vin/rcar-dma.c
index 836d9578b745..3e6ddba52a63 100644
--- a/drivers/media/platform/rcar-vin/rcar-dma.c
+++ b/drivers/media/platform/rcar-vin/rcar-dma.c
@@ -100,8 +100,11 @@
 #define VNMC_INF_YUV8_BT601	(1 << 16)
 #define VNMC_INF_YUV10_BT656	(2 << 16)
 #define VNMC_INF_YUV10_BT601	(3 << 16)
+#define VNMC_INF_RAW8		(4 << 16)
 #define VNMC_INF_YUV16		(5 << 16)
 #define VNMC_INF_RGB888		(6 << 16)
+#define VNMC_INF_RGB666		(7 << 16)
+#define VNMC_INF_MASK		(7 << 16)
 #define VNMC_VUP		(1 << 10)
 #define VNMC_IM_ODD		(0 << 3)
 #define VNMC_IM_ODD_EVEN	(1 << 3)
@@ -910,6 +913,27 @@ static int rvin_setup(struct rvin_dev *vin)
 		interrupts |= VNIE_FOE;
 	}
 
+	/* Check INF bit in VnMR register setting */
+	if (vin->info->model == RCAR_GEN3) {
+		if (vin->mbus_cfg.type == V4L2_MBUS_CSI2) {
+			if (((vnmc & VNMC_INF_MASK) == VNMC_INF_YUV8_BT656) ||
+			    ((vnmc & VNMC_INF_MASK) == VNMC_INF_YUV10_BT656) ||
+			    ((vnmc & VNMC_INF_MASK) == VNMC_INF_YUV16) ||
+			    ((vnmc & VNMC_INF_MASK) == VNMC_INF_RGB666)) {
+				vin_err(vin, "Invalid setting in MIPI CSI2\n");
+
+				return -EINVAL;
+			}
+		} else {
+			if ((vnmc & VNMC_INF_MASK) == VNMC_INF_RAW8) {
+				vin_err(vin,
+					"Invalid setting in Digital Pins\n");
+
+				return -EINVAL;
+			}
+		}
+	}
+
 	/* Ack interrupts */
 	rvin_write(vin, interrupts, VNINTS_REG);
 	/* Enable interrupts */
-- 
2.17.1

